src/material_library.cpp

```diff
@@ -0,0 +1,391 @@
+
+  int file_num_materials = get_length_of_table(filename, datapath);
+  int library_length = material_library.size();
+  for (int i = 0; i < file_num_materials; i++) {
+    pyne::Material mat = pyne::Material();  // from file
+    // read new mat
+    if (nucpath == "") {
+      mat.from_hdf5(filename, datapath, i, protocol);
+    } else {
+      mat.from_hdf5(filename, datapath, nucpath, i, protocol);
+    }
+    (*this).add_material(mat);
+  }
+}
+
+void pyne::MaterialLibrary::merge(pyne::MaterialLibrary mat_lib) {
+  pyne::matname_set mats_to_add = mat_lib.get_keylist();
+  for (auto it = mats_to_add.begin(); it != mats_to_add.end(); it++) {
+    pyne::Material mat = Material(mat_lib.get_material(*it));
+    (*this).add_material(*it, mat);
+  }
+}
+
+void pyne::MaterialLibrary::load_json(Json::Value json) {
+  Json::Value::Members keys = json.getMemberNames();
+  Json::Value::Members::const_iterator ikey = keys.begin();
+  Json::Value::Members::const_iterator ikey_end = keys.end();
+  for (; ikey != ikey_end; ++ikey) {
+    pyne::Material mat = pyne::Material();
+    mat.load_json(json[*ikey]);
+    (*this).add_material(*ikey, Material(mat));
+  }
+}
+
+Json::Value pyne::MaterialLibrary::dump_json() {
+  Json::Value json = Json::Value(Json::objectValue);
+
+  for (auto name : name_order) {
+    json[name] = material_library[name]->dump_json();
+  }
+  return json;
+}
+
+void pyne::MaterialLibrary::from_json(char* fname) {
+  std::string filename(fname);
+  from_json(filename);
+}
+void pyne::MaterialLibrary::from_json(const std::string& filename) {
+  if (!pyne::file_exists(filename)) throw pyne::FileNotFound(filename);
+  std::string s;
+  std::ifstream f(filename.c_str(), std::ios::in | std::ios::binary);
+  f.seekg(0, std::ios::end);
+  s.resize(f.tellg());
+  f.seekg(0, std::ios::beg);
+  f.read(&s[0], s.size());
+  f.close();
+  Json::Reader reader;
+  Json::Value json;
+  reader.parse(s, json);
+  load_json(json);
+}
+
+void pyne::MaterialLibrary::write_json(char* filename) {
+  std::string fname(filename);
+  write_json(fname);
+}
+
+void pyne::MaterialLibrary::write_json(const std::string& filename) {
+  Json::Value json = dump_json();
+  Json::StyledWriter writer;
+  std::string s = writer.write(json);
+  std::ofstream f;
+  f.open(filename.c_str(), std::ios_base::trunc);
+  f << s << "\n";
+  f.close();
+}
+
+void pyne::MaterialLibrary::add_material(pyne::Material mat) {
+  // if exists, get the material name from metadata make one instead
+  std::string mat_name;
+  int mat_number = -1;
+  std::set<int>::iterator mat_numb_it;
+  if (mat.metadata.isMember("mat_number")) {
+    if (Json::intValue <= mat.metadata["mat_number"].type() &&
+        mat.metadata["mat_number"].type() <= Json::realValue ){
+      mat_number = mat.metadata["mat_number"].asInt();
+    } else {
+      mat_number = std::stoi(mat.metadata["mat_number"].asString());
+      mat.metadata["mat_number"] = mat_number;
```

Comment:
I am not sure this is the right way/cleanest way to implement this.
but in some h5m file used in the `nosetests` the `mat_number` is read as a `string`, and the `name` as an `int`

Reply:
To keep previous behavior, I remove the automatic addition of mat_number (when adding material to the MaterialLibrary) if the material does not have it originally